crates/wasmparser/src/readers/core/types.rs

```diff
@@ -595,6 +627,56 @@ pub enum HeapType {
     I31,
 }
 
+impl Matches for HeapType {
+    fn matches<'a, F>(&self, other: &Self, type_at: &F) -> Result<bool>
+    where
+        F: Fn(u32) -> Result<&'a SubType>,
+    {
+        match (self, other) {
+            (HeapType::Indexed(a), HeapType::Indexed(b)) => {
+                Ok(*a == *b || type_at(*a)?.matches(type_at(*b)?, type_at)?)
+            }
+            (HeapType::Indexed(a), HeapType::Func) => match (type_at(*a)?).structural_type {
+                StructuralType::Func(_) => Ok(true),
+                _ => Ok(false),
+            },
+            (HeapType::Indexed(a), HeapType::Array) => match (type_at(*a)?).structural_type {
+                StructuralType::Array(_) => Ok(true),
+                _ => Ok(false),
+            },
+            (HeapType::Indexed(a), HeapType::Struct) => match (type_at(*a)?).structural_type {
+                StructuralType::Struct(_) => Ok(true),
+                _ => Ok(false),
+            },
+            (HeapType::Indexed(a), HeapType::Eq | HeapType::Any) => {
+                match (type_at(*a)?).structural_type {
+                    StructuralType::Array(_) | StructuralType::Struct(_) => Ok(true),
+                    _ => Ok(false),
+                }
+            }
+            (HeapType::Eq, HeapType::Any) => Ok(true),
+            (HeapType::I31 | HeapType::Array | HeapType::Struct, HeapType::Eq | HeapType::Any) => {
+                Ok(true)
+            }
+            (HeapType::None, HeapType::Indexed(a)) => match (type_at(*a)?).structural_type {
+                StructuralType::Array(_) | StructuralType::Struct(_) => Ok(true),
+                _ => Ok(false),
+            },
+            (
+                HeapType::None,
+                HeapType::I31 | HeapType::Eq | HeapType::Any | HeapType::Array | HeapType::Struct,
+            ) => Ok(true),
+            (HeapType::NoExtern, HeapType::Extern) => Ok(true),
+            (HeapType::NoFunc, HeapType::Func) => Ok(true),
+            (HeapType::NoFunc, HeapType::Indexed(a)) => match (type_at(*a)?).structural_type {
+                StructuralType::Func(_) => Ok(true),
+                _ => Ok(false),
+            },
+            (a, b) => Ok(a == b),
```

Comment:
I've mentioned this before, but personally I'm not a fan of catch-all clauses like this. This is very easy to forget to update if a new variant is added because of `#[derive(PartialEq)]` which can lead to further bugs in behavior. Where possible I'd recommend exhaustively matching remaining cases, for example something like:


```
(a @ (HeapType::T1 | HeapType::T2), b) => Ok(a == b),
```

which explicitly documents the list of "catch-all" types being caught here and it's obvious there's no payload so equality can only be identity-based. That way if new types are added in the future this'll generate a compiler error with a non-exhaustive check.

Reply:
> because of `#[derive(PartialEq)]`

Not sure I understand what bad things can happen because of it. Everything else seems clear.

Reply:
Also, updated the code according to your comment.